google/cloud/bigquery/ipython_magics/line_arg_parser/lexer.py

```diff
@@ -0,0 +1,252 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import namedtuple
+from collections import OrderedDict
+import itertools
+import re
+
+import enum
+
+
+Token = namedtuple("Token", ("type_", "lexeme", "pos"))
+StateTransition = namedtuple("StateTransition", ("new_state", "total_offset"))
+
+
+# Token definition order is important, thus an OrderedDict is needed with tightly
+# controlled member definitions (i.e. passed as a sequnce, and *not* via kwargs).
+token_types = OrderedDict(
+    [
+        (
+            "state_1",
+            OrderedDict(
+                [
+                    (
+                        "GOTO_STATE_2",
+                        r"(?P<GOTO_STATE_2>(?=--))",  # double dash - starting the options list
+                    ),
+                    (
+                        "DEST_VAR",
+                        r"(?P<DEST_VAR>[^\d\W]\w*)",  # essentially a Python ID
+                    ),
+                ]
+            ),
+        ),
+        (
+            "state_2",
+            OrderedDict(
+                [
+                    (
+                        "GOTO_STATE_3",
+                        r"(?P<GOTO_STATE_3>(?=--params(?:\s|=|$)))",  # the --params option
+                    ),
+                    ("OPTION_SPEC", r"(?P<OPTION_SPEC>--\w+)"),
+                    ("OPTION_EQ", r"(?P<OPTION_EQ>=)"),
+                    # NOTE: Currently the only valid value for a non "--params" option is
+                    # either a project/table name or an integer (e.g. max_results).
+                    ("OPT_VAL", r"(?P<OPT_VAL>(?:\w|\.|-(?!-))+)"),
```

Comment:
Optionally, we can loosen up the OPT_VAL definition to match just about anything until the next option spec (starting with `--`), or until the next whitespace or end of input.
```py
("OPT_VAL", r"(?P<OPT_VAL>\S+?(?=\s|--|$))")
```
A non-greedy repetition qualifier is needed so that `foo--bar` is correctly recognized as two tokens, namely OPT_VAL, OPT_SPEC, and not as a single OPT_VAL token).

I didn't use this pattern initially, because the lookahead in it somehow triggered catastrophic backtracking and was _way_ too slow, but after a few tweaks to other token definitions, this doesn't appear to be the case anymore.

Let me know if this is something worth changing.

Reply:
I'd say if the only thing needed to bring this additional flexibility is changing the pattern, go for it. If it requires a lot more work, it could be left as is, until usage in the wild lets us know if the flexibility is wanted.

Reply:
Added that, seems to work fine.